orttraining/orttraining/eager/ort_aten.cpp

```diff
@@ -920,6 +920,59 @@ const at::Tensor& resize_(
   return self;
 }
 
+// aten::cat.out(Tensor[] tensors, int dim=0, *, Tensor(a!) out) -> Tensor(a!)
+at::Tensor& cat_out(
+  at::TensorList tensors,
+  int64_t dim,
+  // *,
+  at::Tensor& out) {
+  ORT_LOG_FN(tensors, dim, out);
+
+  assert(tensors.size() > 0);
+  if (
+      std::vector<at::ScalarType> supportedTypes =
+        {at::kBFloat16, at::kBool, at::kByte, at::kDouble, at::kFloat, at::kHalf, at::kInt, at::kLong, at::kShort};
+      !IsSupportedType(tensors, supportedTypes)) {
+    return at::native::call_fallback_fn<
+      &at::native::cpu_fallback,
+      ATEN_OP(cat_out)>::call(tensors, dim, out);
+  }
+  int64_t ndim = tensors[0].dim();
+  assert(ndim != 0);
+  dim = at::maybe_wrap_dim(dim, ndim);
+
+  auto& invoker = GetORTInvoker(tensors[0].device());
+
+  // IntArrayRef isn't writeable, convert to vector.
+  std::vector<int64_t> sizes;
+  for (auto s : tensors[0].sizes())
+    sizes.push_back(s);
+
+  // Calculate the new size of the dimension being concatenated.
+  sizes[dim] = 0;
+  for (auto t : tensors)
```

Comment:
From the [torch.cat](https://pytorch.org/docs/stable/generated/torch.cat.html) documentation - it looks like the input could have empty tensors? I think an empty input tensor could be an issue here.. Maybe just something to check with how default PyTorch implementation handles this case.

Reply:
yes, if the first tensor is truly empty, it could be a problem. I was thinking the tensor at least needed to have the same number of dimensions, even if they were empty, but that might not be 100% accurate

Reply:
I was looking at the following lines from the documentation:

> All tensors must either have the same shape (except in the concatenating dimension) or be empty.

> Non-empty tensors provided must have the same shape, except in the cat dimension.

which I read as empty tensors might not have the same shape as the other tensors.

Reply:
What would/could be the `.sizes()` value of an empty tensor? If all the non-empty tensors had dim=3 but an empty tensor had dim=2 is that okay? I am not worried about this now but something to consider as we move eager forward.